src/include/processor/operator/mask.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <mutex>
+
+#include "storage/store/node_table.h"
+
+namespace kuzu {
+namespace processor {
+
+// Note: All classes in this file are NOT thread-safe.
+struct MaskUtil {
+    static inline common::offset_t getMorselIdx(common::offset_t offset) {
+        return offset >> common::DEFAULT_VECTOR_CAPACITY_LOG_2;
+    }
+};
+
+struct Mask {
+public:
+    explicit Mask(uint64_t size) {
+        data = std::make_unique<uint8_t[]>(size);
+        std::fill(data.get(), data.get() + size, 0);
+    }
+
+    inline void setMask(uint64_t pos, uint8_t maskValue) { data[pos] = maskValue; }
+    inline bool isMasked(uint64_t pos, uint8_t trueMaskVal) { return data[pos] == trueMaskVal; }
+
+private:
+    std::unique_ptr<uint8_t[]> data;
+};
+
+// MaskCollection represents multiple mask on the same domain with AND semantic.
+class MaskCollection {
+public:
+    MaskCollection() : numMasks{0} {}
+
+    inline void init(common::offset_t maxOffset) {
+        if (mask != nullptr) { // MaskCollection might be initialized repeatedly.
+            return;
+        }
+        mask = std::make_unique<Mask>(maxOffset + 1);
+    }
+
+    inline bool isMasked(common::offset_t offset) { return mask->isMasked(offset, numMasks); }
+    // Increment mask value for the given nodeOffset if its current mask value is equal to
+    // the specified `currentMaskValue`.
+    inline void incrementMaskValue(common::offset_t offset, uint8_t currentMaskValue) {
+        if (mask->isMasked(offset, currentMaskValue)) {
+            mask->setMask(offset, currentMaskValue + 1);
+        }
+    }
+
+    inline uint8_t getNumMasks() const { return numMasks; }
+    inline void incrementNumMasks() { numMasks++; }
+
+private:
+    std::unique_ptr<Mask> mask;
+    uint8_t numMasks;
```

Comment:
The term `Mask` here is a bit confusing now, as the `Mask` struct is actually physical representation, while in `numMasks`, "Mask" is a logical concept, we are trying to indicate how many logical masks are passed here and physically stored inside the `std::unique_ptr<Mask> mask;`.

I suggest my changes in the comment under `Mask` struct.